### PR TITLE
Add functions for dragging window separators

### PR DESCRIPTION
**Pull request by dstein64**
#### Description

This exposes the internal functions `win_drag_status_line` and `win_drag_vsep_line` through builtin functions `win_dragstat` and `win_dragvsep`.

#### Motivation

While `:resize` can already be used for resizing windows, it doesn't allow specification for which border will move.

With `win_dragstat` and `win_dragvsep` it's known which border will move.

While `win_dragstat` and `win_dragvsep` can only directly drag a window's right or bottom border, it's still possible to effectively drag a left or top border by dragging the neighbor's right or bottom border:

```vim
:call win_dragvsep(winnr('k'), ...)
:call win_dragvsep(winnr('h'), ...)
```

#### Example Application

The following mappings can be used to drag the current window's borders using control and familiar `hjkl` motions. The `<c-s-...>` mappings for this example depend on a terminal supporting [`modifyOtherKeys`](https://vimhelp.org/map.txt.html#modifyOtherKeys), so that pressing shift can be used to move the left/bottom borders.

```vim
noremap <c-h> :call win_dragvsep(0, -1)<cr>
noremap <c-l> :call win_dragvsep(0, 1)<cr>
noremap <c-k> :call win_dragstat(0, -1)<cr>
noremap <c-j> :call win_dragstat(0, 1)<cr>
noremap <c-s-H> :call win_dragvsep(winnr('h'), -1)<cr>
noremap <c-s-L> :call win_dragvsep(winnr('h'), 1)<cr>
noremap <c-s-K> :call win_dragstat(winnr('k'), -1)<cr>
noremap <c-s-J> :call win_dragstat(winnr('k'), 1)<cr>
```

### CODE REVIEW

**Comment by dstein64** [comment]
#### Alternative Implementation

This could alternatively be implemented as a command. E.g.,

```vim
:{winnr}drags[tat] [+-]N
:{winnr}dragv[sep] [+-]N
```

If the general idea of this PR—exposing the functionality for dragging of window separators without using a mouse—is considered acceptable for merging, **would it be preferable to add this functionality through a builtin function or command?** This PR uses a function, but upon further thought I think I should have implemented this as a command.

**Comment by brammool** [comment]
My first thought was that it would be better to use win_setheight() and win_setwidth() functions, but since this is affecting other windows as well that might also be confusing.  Still "drag" is not really what happens, "move" would be better.
"stat" and "vsep" aren't very clear, "vsep" currently doesn't appear in the help (only in the code).
So how about win_move_statusline() and win_move_separator()?  Those are a bit long.
The closest we already have are the ":resize" commands.  How about win_resize_hor() and win_resize_ver()?

Still, most common use would be to resize a window to a specific size.  Using winheight() and winwidth() one can get the current size, thus if a relative resizing is needed it's a matter of calculating the new size.
Then back to win_setheight() and win_setwidth(), with the documentation explaining that this also affects other windows.
It works like ":resize H" for the height and ":vert resize W" for the width then.

**Comment by dstein64** [comment]
> "So how about win_move_statusline() and win_move_separator()? Those are a bit long."

Those names, `win_move_statusline()` and `win_move_separator()`, sound good to me. I've updated the names and force-pushed.

> "How about win_resize_hor() and win_resize_ver()?"

> "Still, most common use would be to resize a window to a specific size..."

I think that using "resize" could be misleading since the functionality doesn't match how `:resize` works.

The motivation for being able to move the statusline and vertical separator instead of using the existing `:resize` functionality is that it is easier to reason about. An example is in a [question](https://vi.stackexchange.com/questions/22366/how-to-force-which-border-shifts-when-resizing-windows) I asked on [vi.stackexchange.com](https://vi.stackexchange.com) a few years ago. In that particular example, using `win_move_statusline()` could be used to move window 5's top border up, by moving window 3 or 4's status line up (the numbers in the example don't match Vim's window numbering), a movement I could not figure out how to do with existing functionality.

I don't plan to use calls to `win_move_statusline()` and `win_move_separator()` directly through the command line, but rather through mappings and/or a plugin I wrote, [vim-win](https://github.com/dstein64/vim-win), for fine-grained control over window sizing.

File: src/testdir/test_window_cmd.vim [hunk 1]
```diff
@@ -1395,5 +1395,103 @@ func Test_window_minimal_size()
...
+  wincmd =
+  let w = winwidth(0)
+  call assert_false(win_move_separator(-1, 1))
+  call assert_equal(w, winwidth(0))
+  " check that win_move_separator returns false for a popup window
+  let id = popup_create(['hello', 'world'], {})
+  let w = winwidth(id)
+  call assert_false(win_move_separator(id, 1))
+  call assert_equal(w, winwidth(id))
+  call popup_close(id)
```
Comment (dstein64) [inline src/testdir/test_window_cmd.vim:1443]: These lines were added as part of a recent force-push to address uncovered lines in `evalwindow.c` reported by Codecov.

File: src/testdir/test_window_cmd.vim [hunk 2]
```diff
@@ -1395,5 +1395,103 @@ func Test_window_minimal_size()
...
+  wincmd =
+  let h = winheight(0)
+  call assert_false(win_move_statusline(-1, 1))
+  call assert_equal(h, winheight(0))
+  " check that win_move_statusline returns false for a popup window
+  let id = popup_create(['hello', 'world'], {})
+  let h = winheight(id)
+  call assert_false(win_move_statusline(id, 1))
+  call assert_equal(h, winheight(id))
+  call popup_close(id)
```
Comment (dstein64) [inline src/testdir/test_window_cmd.vim:1492]: These lines were added as part of a recent force-push to address uncovered lines in `evalwindow.c` reported by Codecov.

**Comment by brammool** [comment]
In the help there should be some words about other windows also being resized, if they use the same statusline or separator.
Perhaps also mention 'winminheight' and 'winminwidth' so users who wonder why they can't make the window smaller get a hint.

File: runtime/doc/builtin.txt [hunk 3]
```diff
@@ -9562,6 +9564,35 @@ win_id2win({expr})					*win_id2win()*
...
+
+win_move_statusline({nr}, {offset})			*win_move_statusline()*
+		Move window {nr}'s status line (i.e., the bottom border) by
+		{offset} rows, as if being dragged by the mouse. {nr} can be a
+		window number or |window-ID|. A positive {offset} moves down
+		and a negative {offset} moves up. Moving a window's status
+		line will change the height of the window and the height of
+		other windows adjacent to the status line. The magnitude of
+		movement may be smaller than specified (e.g., as a consequence
+		of maintaining 'winminheight'). Returns TRUE if the window can
```
Comment (dstein64) [inline runtime/doc/builtin.txt:9590]: These lines were added as part of a recent force-push in response to a PR review [comment](https://github.com/vim/vim/pull/9486#issuecomment-1008296687).

File: runtime/doc/builtin.txt [hunk 4]
```diff
@@ -9562,6 +9564,35 @@ win_id2win({expr})					*win_id2win()*
...
+win_move_separator({nr}, {offset})			*win_move_separator()*
+		Move window {nr}'s vertical separator (i.e., the right border)
+		by {offset} columns, as if being dragged by the mouse. {nr}
+		can be a window number or |window-ID|. A positive {offset}
+		moves right and a negative {offset} moves left. Moving a
+		window's vertical separator will change the width of the
+		window and the width of other windows adjacent to the vertical
+		separator. The magnitude of movement may be smaller than
+		specified (e.g., as a consequence of maintaining
+		'winminwidth'). Returns TRUE if the window can be found and
```
Comment (dstein64) [inline runtime/doc/builtin.txt:9576]: These lines were added as part of a recent force-push in response to a PR review [comment](https://github.com/vim/vim/pull/9486#issuecomment-1008296687).

**Comment by dstein64** [comment]
Thanks @brammool!

**Pull request closed without merging**